base_layer/core/src/base_node/states/forward_block_sync.rs

```diff
@@ -0,0 +1,278 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    base_node::{
+        states::{ListeningInfo, StateEvent},
+        BaseNodeStateMachine,
+    },
+    blocks::{BlockHash, BlockHeader},
+    chain_storage::{async_db, BlockchainBackend, BlockchainDatabase, ChainMetadata, ChainStorageError},
+    transactions::types::HashOutput,
+};
+use log::*;
+use rand::{rngs::OsRng, Rng};
+use std::{cmp, collections::VecDeque};
+use tari_comms::peer_manager::NodeId;
+use tari_crypto::tari_utilities::{hex::Hex, Hashable};
+
+const LOG_TARGET: &str = "c::bn::states::block_sync";
+
+// The maximum number of retry attempts a node can perform to request a particular block from remote nodes.
+const MAX_HEADER_REQUEST_RETRY_ATTEMPTS: usize = 5;
+const MAX_BLOCK_REQUEST_RETRY_ATTEMPTS: usize = 5;
+const MAX_HEADER_HASHES_TO_SEND: u64 = 128;
+const MAX_BLOCKS_TO_DOWNLOAD: usize = 20;
+
+#[derive(Clone, Debug, PartialEq, Copy)]
+pub struct ForwardBlockSyncInfo;
+
+impl ForwardBlockSyncInfo {
+    pub async fn next_event<B: BlockchainBackend + 'static>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+        sync_peers: &[NodeId],
+    ) -> StateEvent
+    {
+        info!(target: LOG_TARGET, "Synchronizing missing blocks");
+
+        match synchronize_blocks(shared, sync_peers).await {
+            Ok(StateEvent::BlocksSynchronized) => {
+                info!(target: LOG_TARGET, "Block sync state has synchronised");
+                StateEvent::BlocksSynchronized
+            },
+            Ok(state_event) => state_event,
+            Err(e) => StateEvent::FatalError(format!("Synchronizing blocks failed. {}", e)),
+        }
+    }
+}
+
+async fn synchronize_blocks<B: BlockchainBackend + 'static>(
+    shared: &mut BaseNodeStateMachine<B>,
+    sync_nodes: &[NodeId],
+) -> Result<StateEvent, String>
+{
+    let mut sync_nodes = Vec::from(sync_nodes);
+    let tip = shared.db.fetch_tip_header().map_err(|e| e.to_string())?;
+    let mut from_headers = fetch_headers_to_send::<B>(&tip, &shared.db);
+    let mut sync_node = next_sync_node(&mut sync_nodes);
+
+    loop {
+        if sync_node == None {
+            return Err("No more valid nodes to sync to".to_string());
+        }
+        let sync_node_string = sync_node.clone().unwrap();
+
+        info!(
+            target: LOG_TARGET,
+            "Attempting to sync with node:{} asking for headers between heights {} and {}",
+            sync_node_string,
+            from_headers.last().map(|h| h.height).unwrap(),
+            from_headers.first().map(|h| h.height).unwrap(),
+        );
+        match shared
+            .comms
+            .fetch_headers_between(from_headers.iter().map(|h| h.hash()).collect(), None, sync_node.clone())
```

Comment:
If a link is not found, it will return the entire header set from the genesis block to network tip. I believe this will become large as the chain grows.

Reply:
It should return at most `MAX_HEADERS_PER_RESPONSE` headers from the genesis.